src/rust/engine/rule_graph/src/lib.rs

```diff
@@ -435,19 +435,12 @@ impl<'t, R: Rule> GraphMaker<'t, R> {
         // If no candidates were fulfillable, this rule is not fulfillable.
         unfulfillable_diagnostics.push(Diagnostic {
           params: params.clone(),
-          reason: if params.is_empty() {
-            format!(
-              "No rule was available to compute {}. Maybe declare RootRule({})?",
-              dependency_key, product,
-            )
-          } else {
-            format!(
-              "No rule was available to compute {} with parameter type{} {}",
-              dependency_key,
-              if params.len() > 1 { "s" } else { "" },
-              params_str(&params),
-            )
-          },
+          // TODO(#9621)
+          reason: format!(
+            "No rules can compute {}. If you expect to inject this type directly into the rule \
+              graph, rather than deriving it from other rules, then declare RootRule({}).",
```

Comment:
This is VERY good!

Reply:
Credit to @gshuflin! His insight was that we should use domain-specific language for rule authoring errors. My original wording was much vaguer.

Reply:
I think that it might be possible to make an improvement similar to the one that we made for the roots recently here: https://github.com/pantsbuild/pants/commit/700e6ef42956982a471b8bfa772e8b0d7274bb23

In this case, a few lines up we do:
```
for candidate in self.rhs(&params, product) { .. }
```
... to populate the "fulfillable candidates" here. If there are no candidates at all (ie, if `self.rhs` returned no candidates), then there are literally no installed rules that return the relevant type, and so that's an accurate thing to say (in #9136 we went even further, and said "did you mean to install one?" although registering a RootRule as mentioned here might also be a solution).

If, on the other hand, there _were_ rules installed that can return your type, but they were just not fulfillable, this new message you've added might be more accurate... although in that case it's less likely that a RootRule is the right answer: instead, you'd probably want to provide more params: similar to in #9136.

Reply:
I love the error message! So helpful.